src/librustc_trait_selection/traits/structural_match.rs

```diff
@@ -175,7 +180,33 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
                 // for empty array.
                 return false;
             }
-            _ => {
+            ty::Bool
+            | ty::Char
+            | ty::Int(_)
+            | ty::Uint(_)
+            | ty::Float(_)
+            | ty::Str
+            | ty::Never
+            | ty::Error => {
```

Comment:
`Error` is a type error. I don't think that should allow matching?^^

Reply:
We already emitted an error in this case, and I don't want to emit an additionally potentially irrelevant error here.

Reply:
Accepting the wrong kind of type is worse than bad diagnostics... This function here is a key soundness check, the attitude should be "either we are 100% certain this is correct to do under all circumstances, or we reject".

I am not sure how other parts of the type checker deal with `Error`. Maybe `delay_span_bug` would be appropriate.

Reply:
@mark-i-m currently asserts that this is always correct in https://github.com/rust-lang/rust/issues/70551.

Depending on which PR lands first it may still be worth it to add a `delay_span_bug` here, will do once I've finished the test run

Reply:
So, `TyKind::Error` should only be used if we _know_ definitely that an compilation will fail. Unfortunately, there are various hacks that don't use it properly. The last of these is removed in #71930, which has not merged yet. After that, #70551 makes it so that all uses of TyKind::Error will delay a span bug.

@lcnr If this PR touches the exhaustiveness check, then you may want to check that it doesn't implicitly assume #71930...

Reply:
> This function here is a key soundness check, the attitude should be "either we are 100% certain this is correct to do under all circumstances, or we reject".

Actually, I thought it was and it very much sounds like one. But the underlying traits are actually safe, which leaves me rather puzzled as it seems this cannot possibly be used for anything soundness-related then...

Reply:
@mark-i-m 
> If this PR touches the exhaustiveness check,

No it doesn't, it touches "structural match" checking by making it use exhaustive match instead of having a fallback arm.